wasi-tests/src/test/java/com/dylibso/chicory/wasi/WasiPreview1Test.java

```diff
@@ -294,6 +299,134 @@ public void wasiReadLink() throws IOException {
+        int outPtr = inPtr + nsubscriptions * 48;
+        int in = inPtr;
+        memory.writeLong(in, 0x8888_7777_6666_5555L); // userdata
+        memory.writeByte(in + 8, WasiEventType.FD_READ);
+        memory.writeI32(in + 16, 0); // fd
+        int result = wasi.pollOneoff(memory, inPtr, outPtr, nsubscriptions, neventsPtr);
+        assertEquals(WasiErrno.ESUCCESS.value(), result);
+        assertEquals(1, memory.readInt(neventsPtr));
+        assertEquals(0x8888_7777_6666_5555L, memory.readLong(outPtr));
+        assertEquals(WasiErrno.ESUCCESS.value(), memory.readShort(outPtr + 8));
+        assertEquals(WasiEventType.FD_READ, memory.read(outPtr + 10));
+    }
+
+    @Test
+    public void wasiPollOneoffRegularFile() {
+        try (var fs = newJimfs()) {
+            Path root = fs.getPath("test");
+            createDirectory(root);
+            var file = root.resolve("hello.txt");
+            writeString(file, "Hello, World!");
+            try (var wasi = wasiWithDirectory(root.toString(), root)) {
+                var memory = new Memory(new MemoryLimits(1));
+                int fdPtr = 1024;
+                int result = wasi.pathOpen(memory, 3, 0, "hello.txt", 0, 0, 0, 0, fdPtr);
+                assertEquals(WasiErrno.ESUCCESS.value(), result);
+                int fd = memory.readInt(fdPtr);
+                assertEquals(4, fd);
+
+                int nsubscriptions = 2;
+                int neventsPtr = 0;
+                int inPtr = 4;
+                int outPtr = inPtr + nsubscriptions * 48;
+                int in = inPtr;
+                memory.writeLong(in, 0x8888_7777_6666_5555L); // userdata
+                memory.writeByte(in + 8, WasiEventType.FD_READ);
+                memory.writeI32(in + 16, fd);
+                in += 48;
+                memory.writeLong(in, 0xAAAA_BBBB_CCCC_DDDDL); // userdata
+                memory.writeByte(in + 8, WasiEventType.FD_WRITE);
+                memory.writeI32(in + 16, fd);
+                result = wasi.pollOneoff(memory, inPtr, outPtr, nsubscriptions, neventsPtr);
+                assertEquals(WasiErrno.ESUCCESS.value(), result);
+                assertEquals(2, memory.readInt(neventsPtr));
+                int out = outPtr;
+                assertEquals(0x8888_7777_6666_5555L, memory.readLong(out));
+                assertEquals(WasiErrno.ESUCCESS.value(), memory.readShort(out + 8));
+                assertEquals(WasiEventType.FD_READ, memory.read(out + 10));
+                out += 32;
+                assertEquals(0xAAAA_BBBB_CCCC_DDDDL, memory.readLong(out));
+                assertEquals(WasiErrno.ESUCCESS.value(), memory.readShort(out + 8));
+                assertEquals(WasiEventType.FD_WRITE, memory.read(out + 10));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void wasiPollOneoffClockAbstime() {
+        long deadline = System.nanoTime() + MILLISECONDS.toNanos(25);
```

Comment:
Instead of using `System.nanotime` it would be great to use a "mocked" clock.
i.e.:

- set a clock source that can be controlled from the test
- trigger `pollOneOff` and verify that the deadline has not been reached
- increase the clock to be > than the expected deadline
- verify that the deadline is reached

Do you think it's doable?

Reply:
We support mocking for the realtime `Clock` via `WasiOptions`. However, I don't think that simply mocking `System.nanoTime()` is sufficient, as it interacts with the way we use `Thread.sleep()`.

Remember that `poll_oneoff()` waits until there is at least one event. So in order to test that a clock event doesn't trigger, we need a different event to trigger. But then we won't actually sleep, unless we mock `stdin` with something that starts off unreadable and becomes readable later.

I can look at this more if you feel strongly, but I think the testing we have is sufficient, especially for something that likely won't be used for more than a basic sleep function.